src/Features/LanguageServer/Protocol/CustomProtocol/FindUsagesLSPContext.cs

```diff
@@ -0,0 +1,220 @@
+            {
+                if (_definitionToId.TryGetValue(reference.Definition, out var definitionId))
+                {
+                    // Creating a new VSReferenceItem for the reference
+                    var referenceItem = await GenerateVSReferenceItemAsync(
+                        _id, definitionId, _document, _position, reference.SourceSpan,
+                        reference.AdditionalProperties, _metadataAsSourceFileService, definitionText: null,
+                        reference.SymbolUsageInfo, CancellationToken).ConfigureAwait(false);
+
+                    if (referenceItem.Location != null)
+                    {
+                        AddAndReportResultsIfAtMax(referenceItem);
+                        _id++;
+                    }
+                }
+            }
+        }
+
+        private void AddAndReportResultsIfAtMax(VSReferenceItem item)
+        {
+            _resultsChunk.Add(item);
+            if (_resultsChunk.Count < MaxResultsChunkSize)
+            {
+                return;
+            }
+
+            _progress.Report(_resultsChunk.ToArray());
+            _resultsChunk.Clear();
+        }
+
+        private static async Task<LSP.VSReferenceItem> GenerateVSReferenceItemAsync(
+            int id,
+            int? definitionId,
+            Document document,
+            int position,
+            DocumentSpan documentSpan,
+            ImmutableDictionary<string, string> properties,
+            IMetadataAsSourceFileService metadataAsSourceFileService,
+            string? definitionText,
+            SymbolUsageInfo? symbolUsageInfo,
+            CancellationToken cancellationToken)
+        {
+            LSP.Location? location = null;
+
+            // If we have no source span, our location may be in metadata.
+            if (documentSpan == default)
+            {
+                var symbol = await SymbolFinder.FindSymbolAtPositionAsync(document, position, cancellationToken).ConfigureAwait(false);
+                if (symbol != null && symbol.Locations != null && !symbol.Locations.IsEmpty && symbol.Locations.First().IsInMetadata)
+                {
+                    var declarationFile = await metadataAsSourceFileService.GetGeneratedFileAsync(
+                        document.Project, symbol, allowDecompilation: false, cancellationToken).ConfigureAwait(false);
+
+                    var linePosSpan = declarationFile.IdentifierLocation.GetLineSpan().Span;
+                    location = new LSP.Location
+                    {
+                        Uri = new Uri(declarationFile.FilePath),
+                        Range = ProtocolConversions.LinePositionToRange(linePosSpan),
+                    };
+                }
+            }
+            else
+            {
+                location = await ProtocolConversions.DocumentSpanToLocationAsync(documentSpan, cancellationToken).ConfigureAwait(false);
+            }
```

Comment:
this sort of logic is super hard to read the flow of.  so many jumps and dataflow paths the brain has to figure out.